pkg/ccl/changefeedccl/telemetry.go

```diff
@@ -0,0 +1,195 @@
+// Copyright 2023 The Cockroach Authors.
+//
+// Licensed as a CockroachDB Enterprise file under the Cockroach Community
+// License (the "License"); you may not use this file except in compliance with
+// the License. You may obtain a copy of the License at
+//
+//     https://github.com/cockroachdb/cockroach/blob/master/licenses/CCL.txt
+
+package changefeedccl
+
+import (
+	"context"
+	"net/url"
+	"sync/atomic"
+
+	"github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl/changefeedbase"
+	"github.com/cockroachdb/cockroach/pkg/jobs"
+	"github.com/cockroachdb/cockroach/pkg/jobs/jobspb"
+	"github.com/cockroachdb/cockroach/pkg/settings/cluster"
+	"github.com/cockroachdb/cockroach/pkg/util/hlc"
+	"github.com/cockroachdb/cockroach/pkg/util/log"
+	"github.com/cockroachdb/cockroach/pkg/util/log/eventpb"
+	"github.com/cockroachdb/errors"
+)
+
+// A sinkTelemetryDataStore is used to track metrics destined to be
+// logged as telemetry.
+type sinkTelemetryDataStore interface {
+	// incEmittedBytes atomically adds numBytes to the running count of emitted bytes.
+	incEmittedBytes(numBytes int)
+	// resetEmittedBytes atomically resets running count of emitted bytes and
+	// returns the old count.
+	resetEmittedBytes() int
+}
+
+var _ sinkTelemetryDataStore = (*sinkTelemetryData)(nil)
+
+type sinkTelemetryData struct {
+	emittedBytes *atomic.Int64
+}
+
+var noopTelemetryDataStore = (*sinkTelemetryData)(nil)
+
+func makeSinkTelemetryData() *sinkTelemetryData {
+	return &sinkTelemetryData{
+		emittedBytes: &atomic.Int64{},
+	}
+}
+
+// resetEmittedBytes implements the sinkTelemetryDataStore interface.
+func (std *sinkTelemetryData) incEmittedBytes(numBytes int) {
+	if std == nil {
+		return
+	}
+	std.emittedBytes.Add(int64(numBytes))
+}
+
+// resetEmittedBytes implements the sinkTelemetryDataStore interface.
+func (std *sinkTelemetryData) resetEmittedBytes() int {
+	if std == nil {
+		return 0
+	}
+	return int(std.emittedBytes.Swap(0))
+}
+
+type periodicTelemetryLogger struct {
+	sinkTelemetry *sinkTelemetryData
+	lastResolved  hlc.Timestamp
+}
+
+func makePeriodicTelemetryLogger() periodicTelemetryLogger {
+	return periodicTelemetryLogger{
+		sinkTelemetry: makeSinkTelemetryData(),
+	}
+}
+
+func (ptl *periodicTelemetryLogger) maybeLogTelemetry(
+	ctx context.Context,
+	jobID jobspb.JobID,
+	details jobspb.ChangefeedDetails,
+	stmt string,
+	newResolved hlc.Timestamp,
+	s *cluster.Settings,
+	knobs TestingKnobs,
+) {
+	loggingInterval := changefeedbase.ContinuousTelemetryInterval.Get(&s.SV).Nanoseconds()
+	if loggingInterval < 0 {
+		return
+	}
+
+	if newResolved.WallTime-ptl.lastResolved.WallTime >= loggingInterval {
+		changefeedEventDetails := getCommonChangefeedEventDetails(ctx, details, stmt)
```

Comment:
do we need to call this every time? doesn't it stay identical for the duration? (and even if not identical, only few fields in the struct should change -- maybe not a big deal since we intend to call it every 24hrs)..

Btw, 24hrs .. that's a LONG, long time.
Also note, even if you want roughly once a day logging, changefeed may restart (particularly if you say have automatic replanning turned on) -- and this will result in more logging.

Reply:
This is making me question using resolved timestamps for tracking time. 

I think this situation may be possible:
We could have an initial scan on a table with rows having old mvcc timestamps. We could perform a fast initial scan and process data with a wide range of timestamps in just a few minutes, causing many telemetry events to be emitted.